java-support/src/main/java/io/cloudstate/javasupport/crud/CommandContext.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2019 Lightbend Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cloudstate.javasupport.crud;
+
+import io.cloudstate.javasupport.ClientActionContext;
+import io.cloudstate.javasupport.EffectContext;
+
+/**
+ * A CRUD command context.
+ *
+ * <p>Methods annotated with {@link CommandHandler} may take this is a parameter. It allows updating
+ * or deleting the entity state in response to a command, along with forwarding the result to other
+ * entities, and performing side effects on other entities.
+ */
+public interface CommandContext extends CrudContext, ClientActionContext, EffectContext {
+  /**
+   * The current sequence number of state in this entity.
+   *
+   * @return The current sequence number.
+   */
+  long sequenceNumber();
+
+  /**
+   * The name of the command being executed.
+   *
+   * @return The name of the command.
+   */
+  String commandName();
+
+  /**
+   * The id of the command being executed.
+   *
+   * @return The id of the command.
+   */
+  long commandId();
+
+  /**
+   * Update the entity with the new state. The state will be persisted.
+   *
+   * @param state The state to persist.
+   */
+  void update(Object state);
```

Comment:
@ralphlaude Perhaps `updateEntity`? Also, what happens if someone does:

`ctx.deleteEntity()` followed by `ctx.update(…)`?

Reply:
@ralphlaude Should it be possible to update(null)? Also, what happens if the type of the entity denoted by `state` has the wrong type? Is it possible to have a type associated with the CommandContext?

Reply:
@viktorklang 
> 
> Also, what happens if someone does: `ctx.deleteEntity()` followed by `ctx.update(…)`?
the way it is implemented now the User Function will updated accordingly and the CRUD entity will apply only the last command, which is `update` in this case.
The actual implementation only applies the last client action to the CRUD entity.
One solution would be to collect all client actions and update the User Function accordingly and also apply them to the CRUD entity.

Reply:
@viktorklang 
> Should it be possible to update(null)? Also, what happens if the type of the entity denoted by `state` has the wrong type? Is it possible to have a type associated with the CommandContext?

i don't think it should be possible to `update(null)` and it should be prevented with a kind of `RuntimeException`.

If `state` has the wrong type the `UpdateStateHandler` will fail: https://github.com/cloudstateio/cloudstate/blob/c6e162aab5973afbf06796f4a8bfba227b5de1d0/java-support/src/test/scala/io/cloudstate/javasupport/impl/crud/AnnotationBasedCrudSupportSpec.scala#L273
Associating the `CommandContext` with the a type, like this `CommandContext<EntityTpye>`, sounds good and i will investigate that.

Reply:
@viktorklang,
> Also, what happens if someone does: `ctx.deleteEntity()` followed by `ctx.updateEntity(…)`?

in the actual implementation the last CRUD action will win and will be applied. In this case it is `updateEntity`. I am really sure about the most valuable options. I think an valuable option would be to allow the `CommandContext` to collect all CRUD actions and applied them all instead of ignoring some CRUD actions. What do you think?